File: Javascript/30. Substring with Concatenation of All Words.js
/**
 * @param {string} s
 * @param {string[]} words
 * @return {number[]}
 */
var findSubstring = function(s, words) {
    const counterWords = {}
    for (word of words) {
        if (counterWords[word]) {
            counterWords[word]++
        } else {
            counterWords[word] = 1
        }
    }
    
    const lengthWords = words.length
    const lengthElement = words[0].length
    const lengthS = s.length
    const result = []

    for (i=0; i < lengthElement; i++) {
        let counterTemp = {}
        let left = i
        let right = i
        let t = 0
        while (right + lengthElement <= lengthS) {
            let word = s.slice(right, right + lengthElement)
            right += lengthElement
            if (!counterWords[word]) {
                left = right
                counterTemp = {}
                t = 0
                continue
            }

            if (counterTemp[word]) {
                counterTemp[word]++
            } else {
                counterTemp[word] = 1
            }
            t++
            
            while (counterTemp[word] > counterWords[word]) {
                let removeWord = s.slice(left, left + lengthElement)
                left += lengthElement
                counterTemp[removeWord]--
                t--
            }

            if (t == lengthWords) {
                result.push(left)
            }
        }
    }
    console.log(result)
    return result;
};

findSubstring(
    "lingmindraboofooowingdingbarrbarrwingmonkeypoundcake",
    ["fooo","barr","wing","ding","wing"]
)

File: Python/30. Substring with Concatenation of All Words.py
from collections import Counter

class Solution:
    def findSubstring(self, s: str, words: list[str]) -> list[int]:
        counter_words = Counter(words)
        length_words = len(words)
        length_element = len(words[0])
        length_s = len(s)
        result = []

        for i in range(length_element):
            counter_temp = Counter()
            left = right = i
            t = 0
            while right + length_element <= length_s:
                word = s[right : right + length_element]
                right += length_element
                
                if word not in counter_words:
                    left = right
                    counter_temp.clear()
                    t = 0
                    continue

                counter_temp[word] += 1                    
                t += 1

                while counter_temp[word] > counter_words[word]:
                    remove_word = s[left : left + length_element]
                    print(remove_word)
                    left += length_element                 
                    counter_temp[remove_word] -= 1 
                    t -= 1
                if t == length_words:
                    result.append(left)
        
        return result
    

if __name__ == "__main__":
    solution = Solution()
    # For testing
    # print(solution.findSubstring(
    #     s = "wordgoodgoodgoodbestword", 
    #     words = ["word","good","best","word"]
    # ))